public/app/plugins/datasource/prometheus/language_provider.ts

```diff
@@ -111,11 +115,54 @@ export default class PromQlLanguageProvider extends LanguageProvider {
     return defaultValue;
   };
 
+  requestViaWorker(url: string, hasObjectResponse?: false): Promise<string[]>;
+  requestViaWorker(url: string, hasObjectResponse?: true): Promise<PromMetricsMetadata>;
+  requestViaWorker(url: string, hasObjectResponse?: boolean): Promise<string[] | PromMetricsMetadata>;
+  requestViaWorker(url: string, hasObjectResponse = false): Promise<string[] | PromMetricsMetadata> {
+    return new Promise(resolve => {
+      let nodes: string[] | PromMetricsMetadata = hasObjectResponse ? {} : [];
+      const streamWorker = new (StreamJSONResponse as any)() as StreamJSONResponseWorker;
+
+      streamWorker.onmessage = e => {
+        if (e.data === JSON_STREAM_DONE) {
+          streamWorker.terminate();
+          resolve(nodes);
+        }
+
+        if (Array.isArray(nodes)) {
+          nodes = nodes.concat(e.data);
+        } else {
+          Object.assign(nodes, e.data);
+        }
+      };
+
+      streamWorker.onerror = error => {
+        resolve(nodes);
+      };
+
+      streamWorker.postMessage({
+        chunkSize: Math.ceil(this.lookupMetricsThreshold / 20),
+        headers: this.datasource.basicAuth ? { Authorization: this.datasource.basicAuth } : { 'X-Grafana-Org-Id': 1 },
+        limit: this.lookupMetricsThreshold,
+        url: this.datasource.url + url,
+        withCredentials: this.datasource.basicAuth || this.datasource.withCredentials,
+      });
+    });
+  }
+
   start = async (): Promise<any[]> => {
-    this.metrics = await this.request('/api/v1/label/__name__/values', []);
-    this.lookupsDisabled = this.metrics.length > this.lookupMetricsThreshold;
-    this.metricsMetadata = await this.request('/api/v1/metadata', {});
-    this.processHistogramMetrics(this.metrics);
+    const [metrics, metricsMetadata] = await Promise.all([
```

Comment:
Does it make sense to make this cancel-able? We wrap it her ehttps://github.com/grafana/grafana/blob/a4b7c60c2ba5f86ea2ee10b19e24a851e95447c9/public/app/plugins/datasource/prometheus/components/PromQueryField.tsx#L123 with makePromiseCancelable but that just ignores the result. In the worker we check if we are already fetching and so cannot start fetching again. If for some reason the fetching still takes long (slow connection or something) user needs to reload the page cause just navigating somewhere else probably won't work is that right? Maybe we can add something so that this can be cancelled on component unmount which would call worker.terminate()

Reply:
Since `start()` is async, I don't see how it would block anything aside from things that are related to the collection of metrics data.

But... just to be safe, I will try to mock a response that takes forever to resolve and see how this behaves - maybe oboe triggers a timeout anyway. And also how it used to behave. Because maybe it's important for things that depend on it that `start()` simply doesn't finish in such a case.

Reply:
I just looked at the current behavior - it's the same. And it works quite well - the button keeps saying "Loading metrics..." but since it's async, you can of course keep doing everything else, like executing queries and whatnot.

Cleaning up a pending request after a while if the server doesn't respond - I suppose it's cleaner but since it's an improvement over the current behavior and would require quite a bit of effort on my part, I would suggest to do this improvement separately in the future.